File: typescript/src/calisthenics-exercises/no-getters-or-setters.ts
export class BankAccount {
  private _balance: number

  constructor(initialBalance: number = 0) {
    this._balance = initialBalance
  }

  get balance(): number {
    return this._balance
  }

  set balance(value: number) {
    if (value < 0) throw new Error('Negative')
    this._balance = value
  }

  transfer(amount: number, to: BankAccount) {
    this.balance -= amount
    to.balance += amount
  }
}

// Ejemplo de uso:

export function pay(account: BankAccount, amount: number) {
  account.balance = account.balance - amount // lógica externa usando getter/setter
}

File: csharp/src/calisthenics-exercises/NoGettersOrSetters.cs
namespace RefactoringAvanzado.CalisthenicExercises;

public class BankAccount
{
    private double _balance;

    public BankAccount(double initialBalance = 0)
    {
        _balance = initialBalance;
    }

    public double Balance
    {
        get => _balance;
        set
        {
            if (value < 0) throw new Exception("Negative");
            _balance = value;
        }
    }

    public void Transfer(double amount, BankAccount to)
    {
        Balance -= amount;
        to.Balance += amount;
    }
}

public static class BankAccountOperations
{
    public static void Pay(BankAccount account, double amount)
    {
        account.Balance = account.Balance - amount;
    }
}
